### ISSUE TITLE
Test failure under pytest 7.3.0

### ISSUE THREAD

**Issue by bluetech** [pallets/click#2489]
Hi, pytest dev here. We've recently released version 7.3.0. We've had a report that the click test suite fails under 7.3.0. See https://github.com/pytest-dev/pytest/issues/10897 for my analysis. Would love to have your thoughts there.

- Click version: main (1de080b90a82c0ae4e34ba187bc3e945e8a85563)

**Comment by davidism** [comment]
I'm fine with fixing Click. In general, there are a number of issues with how Click's test runner interacts/conflicts with Pytest output handling, PDB, etc. However, I'm not really clear what the issue is, so if you can point to a more self-contained example that demonstrates the issue, or otherwise make a suggestion for a fix, that would be helpful.

**Comment by bluetech** [comment]
The problem is in this line:

https://github.com/pallets/click/blob/1de080b90a82c0ae4e34ba187bc3e945e8a85563/tests/test_arguments.py#L89

which tries to temporarily patch `sys.stdin.encoding` to `utf-8`, but `sys.stdin.encoding` is readonly so it fails.

pytest output capturing (activate by default, turned off with `-s` switch) replaces `sys.stdin` with its own class. Before 7.3.0, that class allowed `sys.stdin.encoding` to be changed, but in 7.3.0 it was changed to match the real `sys.stdin.encoding` to be readonly.

As for fixing the test, I'm didn't try to understand what exactly it's testing, but if it really does want to patch `sys.stdin.encoding`, it's always possible to "force" it like this:

```diff
diff --git a/tests/test_arguments.py b/tests/test_arguments.py
index 735df4b..3395c55 100644
--- a/tests/test_arguments.py
+++ b/tests/test_arguments.py
@@ -1,4 +1,5 @@
 import sys
+from unittest import mock
 
 import pytest
 
@@ -86,9 +87,12 @@ def test_bytes_args(runner, monkeypatch):
         ), "UTF-8 encoded argument should be implicitly converted to Unicode"
 
     # Simulate empty locale environment variables
-    monkeypatch.setattr(sys.stdin, "encoding", "utf-8")
     monkeypatch.setattr(sys, "getfilesystemencoding", lambda: "utf-8")
     monkeypatch.setattr(sys, "getdefaultencoding", lambda: "utf-8")
+    # sys.stdin.encoding is readonly, needs some extra effort to patch.
+    stdin = mock.Mock(wraps=sys.stdin)
+    stdin.encoding = "utf-8"
+    monkeypatch.setattr(sys, "stdin", stdin)
 
     runner.invoke(
         from_bytes,
```

**Comment by saroad2** [comment]
This issue as solved in #2508
